src/DynamoCore/Models/DynamoModel.cs

```diff
@@ -1286,36 +1289,60 @@ public void Paste()
                 AddToSelection(newNote);
             }
 
-            NodeModel start;
-            NodeModel end;
+            ModelBase start;
+            ModelBase end;
             var newConnectors =
                 from c in connectors
 
                 // If the guid is in nodeLookup, then we connect to the new pasted node. Otherwise we
                 // re-connect to the original.
                 let startNode =
-                    nodeLookup.TryGetValue(c.Start.Owner.GUID, out start)
+                    modelLookup.TryGetValue(c.Start.Owner.GUID, out start)
                         ? start
                         : CurrentWorkspace.Nodes.FirstOrDefault(x => x.GUID == c.Start.Owner.GUID)
                 let endNode =
-                    nodeLookup.TryGetValue(c.End.Owner.GUID, out end)
+                    modelLookup.TryGetValue(c.End.Owner.GUID, out end)
                         ? end
                         : CurrentWorkspace.Nodes.FirstOrDefault(x => x.GUID == c.End.Owner.GUID)
 
                 // Don't make a connector if either end is null.
                 where startNode != null && endNode != null
                 select
-                    ConnectorModel.Make(startNode, endNode, c.Start.Index, c.End.Index);
+                    ConnectorModel.Make(startNode as NodeModel, endNode as NodeModel, c.Start.Index, c.End.Index);
 
             createdModels.AddRange(newConnectors);
 
             //Grouping depends on the selected node models. 
             //so adding the group after nodes / notes are added to workspace.
+            //select only those nodes that are part of a group.             
             var newAnnotations = new List<AnnotationModel>();
             foreach (var annotation in annotations)
             {
-                var annotationModel = new AnnotationModel(newNodeModels, newNoteModels) {GUID = Guid.NewGuid()};
-                newAnnotations.Add(annotationModel);              
+                var annotationNodeModel = new List<NodeModel>();
+                var annotationNoteModel = new List<NoteModel>();
+                //checked condition here that supports pasting of multiple groups
+                foreach (var models in annotation.SelectedModels)
+                {
+                    ModelBase mbase;
+                    modelLookup.TryGetValue(models.GUID, out mbase);
+                    if (mbase is NodeModel)
+                    {
+                        annotationNodeModel.Add(mbase as NodeModel);
+                    }
+                    if (mbase is NoteModel)
+                    {
+                        annotationNoteModel.Add(mbase as NoteModel);
```

Comment:
It may seem more wordy, but a faster way to do this is the following:
```
var m = mbase as NodeModel;
if (m != null){
// do something with mbase
} else {
var n = mbase as NoteModel;
if (n != null){
// do something with n
}  
}
```